rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/capabilities/DynamicCapabilities.java

```diff
@@ -0,0 +1,205 @@
+
+                        if (registration == null) {
+                            if (existingRegistration != null) {
+                                // this capability was removed
+                                logger.trace("{} is no longer supported by contributions", method);
+                                unregistrations.add(unregistration(cap));
+                            }
+                            // nothing more to do
+                            continue;
+                        }
+
+                        if (existingRegistration != null) {
+                            if (Objects.deepEquals(registration.getRegisterOptions(), existingRegistration.getRegisterOptions())) {
+                                logger.trace("Options for {} did not change since last registration: {}", method, registration.getRegisterOptions());
+                                continue;
+                            }
+                            logger.trace("Options for {} changed since the previous registration; remove before adding again", method);
+                            unregistrations.add(unregistration(cap));
+                        }
+
+                        logger.trace("Registering dynamic capability {}", method);
+                        registrations.add(registration);
+                    }
+
+                    try {
+                        doRegistrations(registrations, unregistrations);
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                    } catch (ExecutionException e) {
+                        logger.error("Doing registrations failed!", e);
+                    }
+                }
+            }
+        )
+        .exceptionally(e -> {
+            logger.error("Unexpected error occurred while updating dynamic capabilities", e);
+            return null;
+        });
+    }
+
+    private synchronized void doRegistrations(List<Registration> registrations, List<Unregistration> unregistrations) throws InterruptedException, ExecutionException {
+        if (!unregistrations.isEmpty()) {
+            logger.debug("Unregistering dynamic capabilities: {}", unregistrations);
+            client.unregisterCapability(new UnregistrationParams(unregistrations)).get();
+            unregistrations.forEach(u -> currentRegistrations.remove(u.getMethod()));
+        }
+
+        if (!registrations.isEmpty()) {
+            logger.debug("Registering dynamic capabilities: {}", registrations);
+            client.registerCapability(new RegistrationParams(registrations)).get();
+            registrations.forEach(r -> currentRegistrations.put(r.getMethod(), r));
+        }
+    }
+
+    private <T> CompletableFuture<Pair<AbstractDynamicCapability<T>, @Nullable Registration>> maybeRegistration(AbstractDynamicCapability<T> cap, Collection<ILanguageContributions> contribs) {
```

Comment:
I think the name is a bit mislaying I assumed it would check if they "may be  a registration" but it's more `maybeRegister` or `tryRegister`?